WLS JMS connector - Should use Object-Based Security with thin JMS client

New WLS connector initializes InitialContextFactory within different thread than the one which creates destination. This makes
WLS [Thread-Based Security](https://docs.oracle.com/en/middleware/fusion-middleware/weblogic-server/12.2.1.4/jmspg/security.html#GUID-850B39DC-1405-40C6-9959-32BA63E1B733) unusable. We need to switch to [Object-Based Security](https://docs.oracle.com/en/middleware/fusion-middleware/weblogic-server/12.2.1.4/jmspg/security.html#GUID-6F167A65-59F5-4CD5-BEF2-1CE8D9788784).

```
2023.01.12 09:38:05 SEVERE io.helidon.messaging.connectors.jms.JmsConnector Thread[jms-1,5,jms-thread-pool-4]: Error intercepted from channel to-wls
java.util.concurrent.CompletionException: java.lang.RuntimeException: javax.jms.JMSSecurityException: Access denied to resource: type=<jms>, application=osm_oss_jms_module, destinationType=queue, resource=testQueue, action=send
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1770)
	at io.helidon.common.context.Contexts.runInContext(Contexts.java:117)
	at io.helidon.common.context.ContextAwareExecutorImpl.lambda$wrap$7(ContextAwareExecutorImpl.java:154)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.RuntimeException: javax.jms.JMSSecurityException: Access denied to resource: type=<jms>, application=osm_oss_jms_module, destinationType=queue, resource=testQueue, action=send
	at io.helidon.messaging.connectors.wls.ThinClientClassLoader.executeInIsolation(ThinClientClassLoader.java:103)
	at io.helidon.messaging.connectors.wls.WeblogicConnector.consumeAsync(WeblogicConnector.java:165)
	at io.helidon.messaging.connectors.jms.JmsConnector.lambda$consume$17(JmsConnector.java:574)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
	... 5 more
Caused by: javax.jms.JMSSecurityException: Access denied to resource: type=<jms>, application=osm_oss_jms_module, destinationType=queue, resource=testQueue, action=send
	at thinClientClassLoader//weblogic.jms.dispatcher.DispatcherAdapter.convertToJMSExceptionAndThrow(DispatcherAdapter.java:127)
	at thinClientClassLoader//weblogic.jms.dispatcher.DispatcherAdapter.dispatchSyncTran(DispatcherAdapter.java:60)
	at thinClientClassLoader//weblogic.jms.client.JMSProducer.toFEProducer(JMSProducer.java:1559)
	at thinClientClassLoader//weblogic.jms.client.JMSProducer.deliveryInternalUnified(JMSProducer.java:953)
	at thinClientClassLoader//weblogic.jms.client.JMSProducer.sendInternal(JMSProducer.java:618)
	at thinClientClassLoader//weblogic.jms.client.JMSProducer.sendWithListenerUnified(JMSProducer.java:437)
	at thinClientClassLoader//weblogic.jms.client.JMSProducer.sendWithListener(JMSProducer.java:425)
	at thinClientClassLoader//weblogic.jms.client.JMSProducer.send(JMSProducer.java:418)
	at thinClientClassLoader//weblogic.jms.client.WLProducerImpl.send(WLProducerImpl.java:974)
	at io.helidon.messaging.connectors.jms.shim.JakartaMessageProducer.lambda$send$6(JakartaMessageProducer.java:106)
	at io.helidon.messaging.connectors.jms.shim.ShimUtil.run(ShimUtil.java:40)
	at io.helidon.messaging.connectors.jms.shim.JakartaMessageProducer.send(JakartaMessageProducer.java:106)
	at io.helidon.messaging.connectors.jms.JmsConnector.consumeAsync(JmsConnector.java:594)
	at io.helidon.messaging.connectors.wls.WeblogicConnector.lambda$consumeAsync$4(WeblogicConnector.java:165)
	at io.helidon.messaging.connectors.wls.ThinClientClassLoader.executeInIsolation(ThinClientClassLoader.java:101)
	... 8 more

```

### PR's
- [ ] #5852
- [ ] #5854

--- a/docs/mp/reactivemessaging/weblogic.adoc
+++ b/docs/mp/reactivemessaging/weblogic.adoc
@@ -1,6 +1,6 @@
 ///////////////////////////////////////////////////////////////////////////////
 
-    Copyright (c) 2022 Oracle and/or its affiliates.
+    Copyright (c) 2022, 2023 Oracle and/or its affiliates.
 
     Licensed under the Apache License, Version 2.0 (the "License");
     you may not use this file except in compliance with the License.
@@ -42,6 +42,9 @@ WebLogic installation.
 
 WARNING: Avoid placing `wlthint3client.jar` on Helidon classpath, client library location needs to be
 configured and loaded by Helidon messaging connector.
+
+WARNING: Don't forget to start your Helidon app with `--add-opens=java.base/java.io=ALL-UNNAMED` to allow
+wlthint3client use reflection.
 
 include::{rootdir}/includes/dependencies.adoc[]
 

--- a/messaging/connectors/jms/src/main/java/io/helidon/messaging/connectors/jms/ConnectionContext.java
+++ b/messaging/connectors/jms/src/main/java/io/helidon/messaging/connectors/jms/ConnectionContext.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020, 2022 Oracle and/or its affiliates.
+ * Copyright (c) 2020, 2023 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -111,7 +111,7 @@ public class ConnectionContext {
         try {
             return ctx.lookup(jndi);
         } catch (NamingException e) {
-            LOGGER.log(Level.FINE, e, () -> "JNDI lookup of " + jndi + " failed");
+            LOGGER.log(Level.WARNING, e, () -> "JNDI lookup of " + jndi + " failed");
             return null;
         }
     }

--- a/messaging/connectors/wls-jms/src/main/java/io/helidon/messaging/connectors/wls/IsolatedContextFactory.java
+++ b/messaging/connectors/wls-jms/src/main/java/io/helidon/messaging/connectors/wls/IsolatedContextFactory.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022 Oracle and/or its affiliates.
+ * Copyright (c) 2022, 2023 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -18,6 +18,7 @@ package io.helidon.messaging.connectors.wls;
 import java.lang.reflect.Constructor;
 import java.lang.reflect.InvocationTargetException;
 import java.util.Hashtable;
+import java.util.Optional;
 
 import javax.naming.Context;
 import javax.naming.NamingException;
@@ -28,23 +29,26 @@ import javax.naming.spi.InitialContextFactory;
  */
 public class IsolatedContextFactory implements InitialContextFactory {
 
-    private static final String WLS_INIT_CTX_FACTORY = "weblogic.jndi.WLInitialContextFactory";
+    private static final String WLS_INIT_CTX_FACTORY_DEFAULT = "weblogic.jms.WLInitialContextFactory";
 
     @Override
     public Context getInitialContext(Hashtable<?, ?> env) throws NamingException {
         return ThinClientClassLoader.executeInIsolation(() -> {
+            String wlsInitFactoryClass =
+                    Optional.ofNullable((String) env.get("wls-init-ctx-factory"))
+                            .orElse(WLS_INIT_CTX_FACTORY_DEFAULT);
             try {
-                Class<?> wlInitialContextFactory = ThinClientClassLoader.getInstance().loadClass(WLS_INIT_CTX_FACTORY);
+                Class<?> wlInitialContextFactory = ThinClientClassLoader.getInstance().loadClass(wlsInitFactoryClass);
                 Constructor<?> contextFactoryConstructor = wlInitialContextFactory.getConstructor();
                 InitialContextFactory contextFactoryInstance = (InitialContextFactory) contextFactoryConstructor.newInstance();
                 return contextFactoryInstance.getInitialContext(env);
             } catch (ClassNotFoundException e) {
-                throw new RuntimeException("Cannot find " + WLS_INIT_CTX_FACTORY, e);
+                throw new RuntimeException("Cannot find " + wlsInitFactoryClass, e);
             } catch (NoSuchMethodException
                      | InvocationTargetException
                      | InstantiationException
                      | IllegalAccessException e) {
-                throw new RuntimeException("Cannot instantiate " + WLS_INIT_CTX_FACTORY, e);
+                throw new RuntimeException("Cannot instantiate " + wlsInitFactoryClass, e);
             }
         });
     }

--- a/messaging/connectors/wls-jms/src/main/java/io/helidon/messaging/connectors/wls/ThinClientClassLoader.java
+++ b/messaging/connectors/wls-jms/src/main/java/io/helidon/messaging/connectors/wls/ThinClientClassLoader.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022 Oracle and/or its affiliates.
+ * Copyright (c) 2022, 2023 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -57,7 +57,9 @@ class ThinClientClassLoader extends URLClassLoader {
             try {
                 return super.loadClass(name, resolve);
             } catch (ClassNotFoundException e) {
-                LOGGER.log(TRACE, "Cannot load class " + name + " from WLS thin client classloader.", e);
+                LOGGER.log(TRACE, () -> "Cannot load class "
+                        + name
+                        + " from WLS thin client classloader, delegating to ctx classloader.", e);
                 contextClassLoader.loadClass(name);
             }
         }
@@ -94,6 +96,11 @@ class ThinClientClassLoader extends URLClassLoader {
     }
 
     boolean inWlsJar(String name) {
+        // Load jms exceptions from inside the thin jar to avoid deserialization issues
+        if (name.startsWith("javax.jms") && name.endsWith("Exception")) {
+            return true;
+        }
+
         // Load only javax JMS API from outside, so cast works
         return !name.startsWith("javax.jms")
                 && !name.equals(IsolatedContextFactory.class.getName());

